### DISCUSSION TITLE
Separating Recognizer's folder and Namespace

### DISCUSSION THREAD

**Discussion by a-michelis** [antlr/intellij-plugin-v4#567] [Q&A]
Hello! 

I'm using `Rider 2022.2 EAP 4` to create a C# lexer/parser for my project, using the `ANTLR4` plugin. Everything seems to work fine by tweaking the antlr4 settings of each `.g4` file, but i notice a minor problem that i cannot overcome:

let's say the project's name is `MyProject`, located inside `\path\to\Solution\MyProject`. I placed the `.g4` files of lexer and parser at `\path\to\Solution\MyProject\Antlr`. The namespace of the generated classes is `MyProject.Parsing`, with corresponding location `\path\to\Solution\MyProject\Parsing`.

In order to correctly place the generated classes, my config for `.g4` files should be as follows:

**Output directory where all output is generated** :  `\path\to\Solution`
**Location of imported grammars** : `\path\to\Solution\MyProject\Antlr`
**Grammar file encoding** : `utf8`
**Package/namespace for the generated code** : `MyProject.Parsing`
**Language** : `CSharp`

The above configuration works okay, but if the project's folder (`MyProject`) were different from project's base namespace (let's say project's base namespace is`MyProject.Language` with desired output namespace `MyProject.Language.Parsing`), the generator will not place the generated code to the correct location (the code will be placed at `\path\to\Solution\MyProject\Language\Parsing`.

Things are getting even worse if the project's folder has a different name whatsoever (code is not being placed to the project's tree at all).

Is there any way to "decouple" namespace and output directory, or some other workaround?

Thank you in advance.

**Comment by parrt** [comment]
hi. Interesting. I noticed some complexity when I looked at the gen directory thing in the code and couldn't quite figure out exactly what I was doing. The tool does have an option to say where all of the output goes, `-o`, but for backward compatibility reasons tax on the package directory structure UNLESS we use:

```
-Xexact-output-dir  all output goes into -o dir regardless of paths/package
```

When I search in the code it looks like the only place that is used is in building the plugin itself. haha.

I wonder if we shouldn't convert everything to use that exact outfit dirt thing to make it easier to understand we're output goes?
`

**Reply by a-michelis** [reply]
As I haven't interacted with the inner workings of the plugin -or ANTLR4 CLI tool for that matter- , my user experience consists of :

- Using AntlrVSIX on Visual Studio to automate the generation/inclusion process
- Using ANTLR V4 plugin's GUI to automate the generation/inclusion process
- Using ANTLR4 on my projects

I can somewhat understand what you're saying but i'm not even close to being confident enough to give any implementation suggestion whatsoever...

That said, as an end-user, my first time ever opening the configuration window hinted that _Output directory where all output is generated_ were to serve as the **generated code's output directory** and not the **project's or solution's directory**.

I sincerely hope that this feedback can help a bit, as well as help _me_ start diving in the plugin's inner workings.

If I can be of any help for testing or any other reason, I'm up for it!

**Comment by parrt** [comment]
Thanks. next hack session on the plugin, i'll take a look at this. we should simplify.  @bjansen any thoughts?

**Comment by a-michelis** [comment]
Are there any updates on this matter?

My workaround for this, is choosing a completely different path for my generated files and then add them as link.

For a solution with tree like the following:

```
Solution_root
┣━━━Solution.sln
┣━━━src
┃   ┣━━━MyFancyProject
┃   ┃   ┣━━━MyFancyProject.csproj (Namespace: MyFancyProject)
┃   ┃   ┣━━━[...]
┃   ┃
┃   ┣━━━Language
┃   ┃   ┣━━━Language.csproj (Namespace: MyFancyProject.Language)
┃   ┃   ┣━━━[...]
┃   ┃   ┣━━━Antlr
┃   ┃   ┃   ┣━━━Grammar
┃   ┃   ┃   ┃   ┣━━━Lexer.g4
┃   ┃   ┃   ┃   ┣━━━Parser.g4
┃   ┃   ┃   ┣━━━Generated
┃   ┃   ┃   ┃   ┣━━━[This is where i want the Recognizers to be] (Namespace: MyFancyProject.Language.Antlr.Generated)
```

I use a completely different folder `Solution_root/ANTLR4` as root of my generated files. So, my configuration is:

**Output directory where all output is generated** : `Solution_root/ANTLR4`
**Location of imported grammars** : `Solution_root/src/Language/Antlr/Grammar`
**Grammar file encoding** : `utf8`
**Package/namespace for the generated code** : `MyFancyProject.Language.Antlr.Generated`
**Language** : `CSharp`

After that, i add the following to my `Language.csproj` :

```xaml
<Project>
    <!--- [...] --->
    <ItemGroup>
        <Compile Include="..\..\ANTLR4\MyFancyProject\Language\Antlr\Generated\*.cs">
            <Link>Antlr\Generated\%(Filename)%(Extension)</Link>
        </Compile>
    </ItemGroup>
</Project>
```

![Result]( http://rate-it.xyz/download.png )

**Comment by parrt** [comment]
Sorry but I've been unable to get back to this project
